File: dashboard/api/schemas/streaming.py
"""WebSocket message schemas for real-time agent progress streaming."""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class WSMessageType(str, Enum):
    AGENT_STATUS = "agent_status"
    REPORT_CHUNK = "report_chunk"
    STATS = "stats"
    MESSAGE = "message"
    TOOL_CALL = "tool_call"
    COMPLETION = "completion"
    ERROR = "error"
    PIPELINE_STAGE = "pipeline_stage"


class AgentStatus(str, Enum):
    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    ERROR = "error"


class AgentStatusPayload(BaseModel):
    agent: str = Field(..., description="Agent name, e.g. 'Market Analyst'")
    status: AgentStatus


class ReportChunkPayload(BaseModel):
    section: str = Field(..., description="Report section key, e.g. 'market_report'")
    content: str


class StatsPayload(BaseModel):
    llm_calls: int
    tool_calls: int
    tokens_in: int
    tokens_out: int
    elapsed_seconds: float


class CompletionPayload(BaseModel):
    final_decision: str
    rating: str
    ticker: str
    date: str


class ErrorPayload(BaseModel):
    message: str
    agent: Optional[str] = None


class WSMessage(BaseModel):
    type: WSMessageType
    timestamp: str = Field(default_factory=lambda: datetime.utcnow().isoformat())
    payload: dict

File: dashboard/frontend/src/types/streaming.ts
export type AgentStatus = 'pending' | 'in_progress' | 'completed' | 'error'

export interface AgentStatusPayload {
  agent: string
  status: AgentStatus
}

export interface ReportChunkPayload {
  section: string
  content: string
}

export interface StatsPayload {
  llm_calls: number
  tool_calls: number
  tokens_in: number
  tokens_out: number
  elapsed_seconds: number
}

export interface CompletionPayload {
  final_decision: string
  rating: string
  ticker: string
  date: string
}

export interface ErrorPayload {
  message: string
  agent?: string
}

export interface HumanReviewPayload {
  review_point: 'research_manager' | 'portfolio_manager'
  ticker: string
  data: Record<string, unknown>
}

export type WSMessageType =
  | 'agent_status'
  | 'report_chunk'
  | 'stats'
  | 'message'
  | 'tool_call'
  | 'completion'
  | 'error'
  | 'pipeline_stage'
  | 'human_review_required'

export interface WSMessage {
  type: WSMessageType
  timestamp: string
  payload: AgentStatusPayload | ReportChunkPayload | StatsPayload | CompletionPayload | ErrorPayload | HumanReviewPayload | Record<string, unknown>
}
